core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/sync/BufferingContentStreamProvider.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.internal.sync;
+
+import static software.amazon.awssdk.utils.FunctionalUtils.invokeSafely;
+
+import java.io.BufferedInputStream;
+import java.io.InputStream;
+import software.amazon.awssdk.annotations.NotThreadSafe;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.io.ReleasableInputStream;
+import software.amazon.awssdk.http.ContentStreamProvider;
+import software.amazon.awssdk.utils.IoUtils;
+
+@SdkInternalApi
+@NotThreadSafe
+public final class BufferingContentStreamProvider implements ContentStreamProvider {
+    private final ContentStreamProvider delegate;
+    private InputStream bufferedStream;
+
+    public BufferingContentStreamProvider(ContentStreamProvider delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public InputStream newStream() {
+        if (bufferedStream == null) {
+            // Set up to buffer the stream from the wrapped content provider, so that we only call `newStream()` at most once.
+            // This logic is similar to what we do for RequestBody.fromInputStream()
+            // - We mark the stream (here, Integer.MAX_VALUE rather than 128KiB)
+            // - Make it non-closeable since the SDK will close the current stream before calling `newStream`
+            InputStream delegateStream = delegate.newStream();
+            bufferedStream = new BufferedInputStream(delegateStream);
```

Comment:
When will this stream be closed?

Reply:
Right now, never.

 - One thing we can do is intercept when `delegateStream` reads pas EOF and then close it, so we just have the buffered data in memory.

After that, we can only rely on GC.